### PR TITLE
[SPARK-14833][SQL][STREAMING][TEST] Refactor StreamTests to test for source fault-tolerance correctly.

### PR DESCRIPTION
**Pull request by tdas**
## What changes were proposed in this pull request?

Current StreamTest allows testing of a streaming Dataset generated explicitly wraps a source. This is different from the actual production code path where the source object is dynamically created through a DataSource object every time a query is started. So all the fault-tolerance testing in FileSourceSuite and FileSourceStressSuite is not really testing the actual code path as they are just reusing the FileStreamSource object. 

This PR fixes StreamTest and the FileSource***Suite to test this correctly. Instead of maintaining a mapping of source --> expected offset in StreamTest (which requires reuse of source object), it now maintains a mapping of source index --> offset, so that it is independent of the source object.

Summary of changes
- StreamTest refactored to keep track of offset by source index instead of source
- AddData, AddTextData and AddParquetData updated to find the FileStreamSource object from an active query, so that it can work with sources generated when query is started.
- Refactored unit tests in FileSource***Suite to test using DataFrame/Dataset generated with public, rather than reusing the same FileStreamSource. This correctly tests fault tolerance.

The refactoring changed a lot of indents in FileSourceSuite, so its recommended to hide whitespace changes with this - https://github.com/apache/spark/pull/12592/files?w=1
## How was this patch tested?

Refactored unit tests.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 1]
```diff
@@ -369,6 +369,7 @@ class StreamExecution(
   def awaitOffset(source: Source, newOffset: Offset): Unit = {
     def notDone = {
       val localCommittedOffsets = committedOffsets
+      logInfo(s"Current offset for $source = " + localCommittedOffsets.get(source))
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala]: note to self: remove this

File: sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala [hunk 2]
```diff
@@ -67,12 +67,6 @@ import org.apache.spark.util.Utils
  */
 trait StreamTest extends QueryTest with Timeouts {
 
-  implicit class RichSource(s: Source) {
-    def toDF(): DataFrame = Dataset.ofRows(sqlContext, StreamingExecutionRelation(s))
-
-    def toDS[A: Encoder](): Dataset[A] = Dataset(sqlContext, StreamingExecutionRelation(s))
-  }
-
   /** How long to wait for an active stream to catch up when checking a result. */
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala]: Removed the above lines to make sure that in future we do not use this pattern of creating DF/DS on source for testing.

File: sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala [hunk 3]
```diff
@@ -95,20 +89,18 @@ trait StreamTest extends QueryTest with Timeouts {
...
   /** A trait that can be extended when testing other sources. */
   trait AddData extends StreamAction {
-    def source: Source
-
     /**
      * Called to trigger adding the data.  Should return the offset that will denote when this
      * new data has been processed.
      */
-    def addData(): Offset
+    def addData(query: Option[StreamExecution]): (Source, Offset)
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala]: note to self: update docs.

**Comment by tdas** [comment]
@zsxwing

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 4]
```diff
@@ -375,10 +406,10 @@ class FileStreamSourceStressTestSuite extends FileStreamSourceTest with SharedSQ
     val src = Utils.createTempDir(namePrefix = "streaming.src")
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: nit: This test can use `withTempDirs`

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 5]
```diff
@@ -205,165 +262,139 @@ class FileStreamSourceSuite extends FileStreamSourceTest with SharedSQLContext {
...
+    withTempDirs { case (src, tmp) =>
 
-    val textSource = createFileStreamSource("json", src.getCanonicalPath)
+      // Add a file so that we can infer its schema
+      stringToFile(new File(src, "existing"), "{'c': 'drop1'}\n{'c': 'keep2'}\n{'c': 'keep3'}")
 
-    // FileStreamSource should infer the column "c"
-    val filtered = textSource.toDF().filter($"c" contains "keep")
+      val fileStream = createFileStream("json", src.getCanonicalPath)
+      require(fileStream.schema === StructType(Seq(StructField("c", StringType))))
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: nit: require -> assert

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala [hunk 6]
```diff
@@ -19,72 +19,129 @@ package org.apache.spark.sql.streaming
...
       } else {
         sqlContext.read.format(format)
       }
     reader.stream(path)
-      .queryExecution.analyzed
+  }
+
+  protected def getSourceFromFileStream(df: DataFrame): FileStreamSource = {
+    val checkpointLocation = Utils.createTempDir(namePrefix = "streaming.metadata").getCanonicalPath
+    Utils.createTempDir(namePrefix = "streaming.metadata").getCanonicalPath
```
Comment (zsxwing) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala]: nit: unused line.

**Comment by zsxwing** [comment]
LGTM except some nits and the style issue.

**Comment by zsxwing** [comment]
LGTM. Thanks, merging to master.

**Pull request closed without merging**